addons/mail/static/src/core/common/data_model.js

```diff
@@ -0,0 +1,64 @@
+import { Record } from "@mail/core/common/record";
+
+import { Deferred } from "@web/core/utils/concurrency";
+
+/**
+ * This class represents a specific and unique request coming from the client to the server.
+ * It is useful when batching requests together to determine which data to return for a specific
+ * request. It basically does the un-batching by referencing the given id. The rest of the data
+ * being grouped together in the store, flattened and with no duplicate records/fields, no matter
+ * how many requests are returning the same data.
+ * Instances of the class are created by the fetch method on each call, and they are deleted once
+ * they are resolved with their data. This class should not be used directly under typical use.
+ */
+export class Data extends Record {
+    static id = "id";
+    static _nextId = 0;
+
+    static createRequest() {
+        return this.insert({ id: ++this._nextId });
+    }
+
+    /** @type {number} */
+    id;
+    /**
+     * When set to true, this data request is resolved as soon as its RPC returns, even if there was
+     * no actual data for this request inside it. This is useful for fetch request that only fills
+     * the store but does not need to wait any particular value.
+     */
+    _autoResolve = false;
+    /**
+     * Promise that is resolved with the data when the data request is complete.
+     */
+    _resultDef = new Deferred();
+    /**
+     * When set to true, resolves this data request with the current values of the fields as data,
+     * and then deletes the data request.
+     *
+     * @type {boolean}
+     */
+    _resolve = Record.attr(undefined, {
+        /** @this {import("models").Data} */
+        onUpdate() {
+            if (this._resolve) {
+                this._resultDef.resolve({ ...this });
```

Comment:
Note: copy to keep ref after delete. We might want to return only actual fields instead?
But it seems to work just fine (with a little bit of extra noise in the result).